modules/placement-driver/src/main/java/org/apache/ignite/internal/raft/client/TopologyAwareRaftGroupService.java

```diff
@@ -0,0 +1,443 @@
+
+    /** Raft message factory. */
+    private final RaftMessagesFactory factory;
+
+    /** Cluster service. */
+    private final ClusterService clusterService;
+
+    /** RPC RAFT client. */
+    private final RaftGroupService raftClient;
+
+    /** Logical topology service. */
+    private final LogicalTopologyService logicalTopologyService;
+
+    /** Leader election handler. */
+    private final ServerEventHandler serverEventHandler;
+
+    /** Executor to invoke RPC requests. */
+    private final ScheduledExecutorService executor;
+
+    /** RAFT configuration. */
+    private final RaftConfiguration raftConfiguration;
+
+
+    /**
+     * The constructor.
+     *
+     * @param cluster Cluster service.
+     * @param factory Message factory.
+     * @param executor RPC executor.
+     * @param raftClient RPC RAFT client.
+     * @param logicalTopologyService Logical topology.
+     */
+    private TopologyAwareRaftGroupService(
+            ClusterService cluster,
+            RaftMessagesFactory factory,
+            ScheduledExecutorService executor,
+            RaftConfiguration raftConfiguration,
+            RaftGroupService raftClient,
+            LogicalTopologyService logicalTopologyService
+    ) {
+        this.clusterService = cluster;
+        this.factory = factory;
+        this.executor = executor;
+        this.raftConfiguration = raftConfiguration;
+        this.raftClient = raftClient;
+        this.logicalTopologyService = logicalTopologyService;
+        this.serverEventHandler = new ServerEventHandler();
+
+        cluster.messagingService().addMessageHandler(RaftMessageGroup.class, (message, senderConsistentId, correlationId) -> {
+            if (message instanceof SubscriptionLeaderChangeResponse) {
+                var msg = (SubscriptionLeaderChangeResponse) message;
+
+                serverEventHandler.onLeaderElected(clusterService.topologyService().getByConsistentId(senderConsistentId), msg.term());
+            }
+        });
+
+        logicalTopologyService.addEventListener(new LogicalTopologyEventListener() {
+            @Override
+            public void onAppeared(ClusterNode appearedNode, LogicalTopologySnapshot newTopology) {
+                for (Peer peer : peers()) {
+                    if (serverEventHandler.isSubscribed() && appearedNode.name().equals(peer.consistentId())) {
+                        LOG.info("New peer will be sending a leader elected notification [grpId={}, consistentId={}]", groupId(),
+                                peer.consistentId());
+
+                        sendSubscribeMessage(appearedNode, TopologyAwareRaftGroupService.this.factory.subscriptionLeaderChangeRequest()
+                                .groupId(groupId())
+                                .subscribe(true)
+                                .build())
+                                .thenComposeAsync(couldLeaderChange -> {
+                                    if (couldLeaderChange) {
```

Comment:
maybe `topologyChanged`?

Reply:
There is no obligation the topology is changed. I'm going to change the name to more obvious.
For example: `subscribed`